perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -4508,21 +4509,52 @@ public Object getAttributesModule(PerunSession sess, AttributeDefinition attribu
 	}
 
 	@Override
-	public void initAttributeModules(ServiceLoader<AttributesModuleImplApi> modules) {
-		for (AttributesModuleImplApi module : modules) {
-			attributesModulesMap.put(module.getClass().getName(), module);
-			log.debug("Module {} loaded.", module.getClass().getSimpleName());
+	public AttributesModuleImplApi getUninitiatedAttributesModule(PerunSession sess, AttributeDefinition attributeDefinition) {
+
+		// core attributes doesn't have modules
+		if (isCoreAttribute(sess, attributeDefinition)) return null;
+
+		String moduleName = attributeNameToModuleName(attributeDefinition.getNamespace() + ":" + attributeDefinition.getFriendlyName());
+		AttributesModuleImplApi module = uninitializedAttributesModulesMap.get(moduleName);
+
+		if (module == null) {
+			moduleName = attributeNameToModuleName(attributeDefinition.getNamespace() + ":" + attributeDefinition.getBaseFriendlyName());
+			module = uninitializedAttributesModulesMap.get(moduleName);
 		}
+
+		return module;
 	}
 
 	@Override
-	public void registerAttributeModules(ServiceLoader<AttributesModuleImplApi> modules) {
+	public void initAndRegisterAttributeModules(PerunSession sess, ServiceLoader<AttributesModuleImplApi> modules, Set<AttributeDefinition> allAttributesDef) {
 		for (AttributesModuleImplApi module : modules) {
-			Auditer.registerAttributeModule(module);
-			log.debug("Module {} was registered for audit message listening.", module.getClass().getName());
+			uninitializedAttributesModulesMap.put(module.getClass().getName(), module);
+		}
+
+		for (AttributeDefinition attributeDefinition : allAttributesDef) {
+			AttributesModuleImplApi module = getUninitiatedAttributesModule(sess, attributeDefinition);
+
+			if (module != null) {
+				attributesModulesMap.put(module.getClass().getName(), module);
```

Comment:
I would rather call ``initAttributeModule()``, which you added below. Also you can move logging messages in those methods.

Reply:
Fixed.